### PR TITLE
Breaking: Change `context.report` to take an `options` object 

### PR DESCRIPTION
**Pull request by antross**
## Pull request checklist

Make sure you:

- [x] Signed the [Contributor License Agreement](https://cla.js.foundation/webhintio/hint)
- [x] Followed the [commit message guidelines](https://webhint.io/docs/contributor-guide/getting-started/pull-requests/#commit-messages)

For non-trivial changes, please make sure you also:

- [x] Added/Updated related documentation.
- [x] Added/Updated related tests.

## Short description of the change(s)

Fix webhintio/rfcs#47.

Now `report` only has two required parameters, a string `resource`
and a string `message`. All other parameters are now optionally
passed as part of a third `options` parameter of type
`ReportOptions`.

Old code like:
```ts
await context.report(
    resource,
    null,
    message,
    undefined,
    location
);
```

Can now be written as:
```ts
await context.report( resource, message, { location });
```

### CODE REVIEW

**Comment by antross** [comment]
Marking `work-in-progress` as I still need to add official documentation for the full set of options available on `report`.

**Comment by antross** [comment]
@molant although we haven't touched the package in 8 months this is also now failing when I run the test against master. 

Digging into the [amphtml-validator npm package](https://github.com/ampproject/amphtml/tree/master/validator/nodejs/) on GitHub, it looks like most of the logic is contained in a `validator.js` file that is dynamically fetched from `https://cdn.ampproject.org/v0/validator.js` instead of being included in the package. This file is built from the core [amphtml-validator folder](https://github.com/ampproject/amphtml/tree/master/validator) which is not published on npm. Checking the HTTP headers of this online, it has a last-modified header for yesterday: Wed, 17 Oct 2018 19:15:00 GMT.

For now I'm going to update the test (in a separate PR so we can merge it quickly as this affects all our branches due to being dynamically fetched), but I'd like your input on what we want to do with this test going forward. We have the option of passing our own `validator.js` to the call to `getInstance` so we could keep a downloaded copy in-box to avoid being surprised by future updates. That said, we'd need to manually update this so it may not be worth it unless we feel we need to eliminate the extra service call.

**Comment by antross** [comment]
@molant, just saw you found and updated that last night as I rebased master - still curious about your thoughts for this package going forward.

**Comment by molant** [comment]
> still curious about your thoughts for the hint-amp-validator package going forward.

We haven't had any problem in the last few months so I would do anything right now. That said, I think we should open an issue with low priority with what you propose in case someone wants to help with it. If it's documented well enough it could be a good first contribution.

**Comment by molant** [comment]
@antross you have some linting issue in the `apple-touch-icon` hint that fails the build. Can you please fix it?

**Comment by antross** [comment]
> you have some linting issue in the apple-touch-icon hint that fails the build. Can you please fix it?

🤦‍♂️ - on it

**Comment by antross** [comment]
@alrra This is now up-to-date when you have time to take a look. CI is passing on the node versions for which canvas-prebuilt currently works.

File: packages/hint/docs/contributor-guide/how-to/hint.md [hunk 1]
```diff
@@ -94,15 +94,29 @@ with the website and report errors.
...
 ```
 
 * `context.report()` is an asynchronous method, you should always `await`.
 * `resource` is the URL of what is being analyzed (HTML, JS, CSS, manifest,
   etc.)
-* `element` is the `IAsyncHTMLElement` that triggered the problem. Not always
-  necessary. In the case of an image, script, style, it will be an `img`,
-  `script`, `link`, etc.
+* `message` is the text to show to the user about the problem.
+* `element` is an optional `IAsyncHTMLElement` where the issue was found
```
Comment (molant) [inline packages/hint/docs/contributor-guide/how-to/hint.md]: I think this is a bit confusing. The third parameter is the `options` object that can have an `element` property, right? I think we should make it clearer. Right now this splits the description between `element` and `options`. What about something like:

```ts
await context.report(resource, message, options);
```
* `context.report()` is an asynchronous method, you should always `await`.
* `resource` is the URL of what is being analyzed (HTML, JS, CSS, manifest,
  etc.)
* `message` is the text to show to the user about the problem.
* options is an (optional) object that can contain the following:
  * `element` is an optional `IAsyncHTMLElement` where the issue was found ...
  * `codeSnippet` is a string of source code to display (defaults to the
  `outerHTML` of `element`).
  * ...

**Comment by alrra** [comment]
@antross Can you take care of the conflicts? Thanks!

**Comment by antross** [comment]
Yep, I'll have them fixed shortly.

**Pull request closed without merging**